File: Aoc2025/Day2/BenchmarkSolutions.py
import time
import math as m

# DUMMY DATA FOR BENCHMARKING
ranges = [(100_000, 200_000), (900_000_000, 900_050_000)]

## Optimized Mathematical Approach Function
def solve_math(ranges):
    count = 0
    for r in ranges:
        for id in range(r[0], r[1] +1):
            # Get number of digits
            d = int(m.log10(id) + 1)
            for mask in get_mask(d):  # Retrieve from cache
                if id % mask == 0:
                    count += 1
                    break

    return count

mask_cache = {}
## Caching Method for Optimizing Math Approach
def get_mask(d):
    if d in mask_cache:
        return mask_cache[d]

    masks = []
    for l in range(1, (d//2) + 1):
        if d % l == 0:
            mask = (10 ** d - 1) // (10 ** l - 1)
            masks.append(mask)
    
    mask_cache[d] = masks
    return masks

## Optimized String Doubling Approach Function
def solve_string(ranges):
    count = 0
    for r in ranges:
        for id in range(r[0], r[1] +1):
            s = str(id)
            d = s + s
            if s in d[1:-1]:
                count += 1
    return count

# --- Bencmark Math Approach ---
start_time = time.perf_counter()
math_count = solve_math(ranges)
end_time = time.perf_counter()
print(f"Mathematical Approach:   {end_time - start_time:.4f} seconds (Found: {math_count})")

# --- Benchmark String Doubling Approach ---
start_time = time.perf_counter()
string_count = solve_string(ranges)
end_time = time.perf_counter()
print(f"String Doubling Approach: {end_time - start_time:.4f} seconds (Found: {string_count})")

File: Aoc2025/Day2/BenchmarkSolutions.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <map>
#include <chrono>

using namespace std;

// --- HELPER: Integer Power Function ---
long long ipow(long long base, int exp) {
    long long res = 1;
    for (int i = 0; i < exp; i++) res *= base;
    return res;
}

// --- MATH APPROACH ---
map<int, vector<long long>> mask_cache;

vector<long long> get_masks(int d) {
    if (mask_cache.count(d)) {
        return mask_cache[d];
    }

    vector <long long> masks;
    
for (int l = 1; l <= d / 2; l++) {
        if (d % l == 0) {
            long long numerator = ipow(10, d) -1;
            long long denominator = ipow(10, l) -1;
            masks.push_back(numerator / denominator);
        }
    }

    mask_cache[d] = masks;
    return masks;
}

long long solve_math(const vector<pair<long long, long long>> & ranges){
    long long count = 0;
    for (const auto& r : ranges) {
        for (long long id = r.first; id <= r.second; id++) {
            int d = (int)log10(id) + 1;
            
            const vector<long long>& masks = get_masks(d);

            for (long long mask: masks) {
                if (id % mask == 0){
                    count ++;
                    break;
                }
            }
        }
    }
    return count;
}

// --- STRING APPROACH ---
long long solve_string(const vector<pair<long long, long long>> & ranges) {
    long long count = 0;
    for (const auto& r : ranges) {
        for (long long id = r.first; id <= r.second; id++) {
            string s = to_string(id);

            string doubled = s + s;

            string searchArea = doubled.substr(1, doubled.length() -2);

            if (searchArea.find(s) != string::npos) {
                count ++;
            }
        }
    }
    return count;
}

int main() {
    // Dummy Data
    vector<pair<long long, long long>> ranges = { {100000, 200000},
     {500000000, 1000000000} 
    };

    cout << "Starting C++ Benchmark..." << endl;

    // --- BENCHMARK MATH ---
    auto start = chrono::high_resolution_clock::now();
    long long math_count = solve_math(ranges);
    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> diff = end - start;
    cout << "Math Method: " << diff.count() << "s ( Found:" << math_count << " )" << endl;

    // --- Benchmark String ---
    start = chrono::high_resolution_clock::now();
    long long string_count = solve_string(ranges);
    end = chrono::high_resolution_clock::now();
    diff = end - start;
    cout << "String Method: " << diff.count() << " s (Found: " << string_count << ")" << endl;

    return 0;
}